plugin/host.go

```
@@ -0,0 +1,97 @@
+)
+
+func instantiateHostScheduler(ctx context.Context, runtime wazero.Runtime) (wazeroapi.Module, error) {
+	return runtime.NewHostModuleBuilder(k8sScheduler).
+		NewFunctionBuilder().
+		WithGoModuleFunction(wazeroapi.GoModuleFunc(k8sSchedulerStatusReasonFn), []wazeroapi.ValueType{i32, i32}, []wazeroapi.ValueType{}).
+		WithParameterNames("ptr", "size").Export(k8sSchedulerStatusReason).
+		Instantiate(ctx)
+}
+
+func instantiateHostApi(ctx context.Context, runtime wazero.Runtime) (wazeroapi.Module, error) {
+	return runtime.NewHostModuleBuilder(k8sApi).
+		NewFunctionBuilder().
+		WithGoModuleFunction(wazeroapi.GoModuleFunc(k8sApiNodeInfoNodeFn), []wazeroapi.ValueType{i32, i32}, []wazeroapi.ValueType{i32}).
+		WithParameterNames("buf", "buf_limit").Export(k8sApiNodeInfoNode).
+		NewFunctionBuilder().
+		WithGoModuleFunction(wazeroapi.GoModuleFunc(k8sApiPodSpecFn), []wazeroapi.ValueType{i32, i32}, []wazeroapi.ValueType{i32}).
+		WithParameterNames("buf", "buf_limit").Export(k8sApiPodSpec).
+		Instantiate(ctx)
+}
+
+// filterArgsKey is a context.Context value associated with a filterArgs
+// pointer to the current request.
+type filterArgsKey struct{}
+
+type filterArgs struct {
+	pod      *v1.Pod
+	nodeInfo *framework.NodeInfo
+	// reason is a field to avoid compiler-specific malloc/free functions, and
+	// to avoid having to deal with out-params because TinyGo only supports a
+	// single result.
+	reason string
+}
+
+func filterArgsFromContext(ctx context.Context) *filterArgs {
+	return ctx.Value(filterArgsKey{}).(*filterArgs)
+}
+
+// k8sSchedulerStatusReasonFn is a function used by the wasm guest to set the
+// framework.Status reason.
+func k8sSchedulerStatusReasonFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
+	ptr := uint32(stack[0])
+	size := bufLimit(stack[1])
+
+	var reason string
+	if b, ok := mod.Memory().Read(ptr, size); !ok {
+		// don't panic if we can't read the message.
+		reason = "BUG: out of memory reading message"
```

Just curious, what happen if the host function panices?

---

Leave my curiosity aside. 
In such error cases, we want to return an error from `Filter()` to the scheduling framework eventually.

---

so any host function panic is returned as an error from wazero api.Function call. So this means the scheduling framework sees it as an error return.